openedx_blockstore_relay/block_serializer.py

```diff
@@ -0,0 +1,157 @@
+    into Blockstore, we assume that each usage is unique, don't generate a usage
+    key, and create a new "definition key" from the original usage key.
+
+    So modulestore usage key
+        block-v1:A+B+C+type@html+block@introduction
+    will become Blockstore definition key
+        html/introduction
+    """
+    block_type = usage_key.block_type
+    if block_type == 'vertical':
+        # We transform <vertical> to <unit>
+        block_type = "unit"
+    return block_type + "/" + usage_key.block_id
+
+
+class XBlockSerializer(object):
+    """
+    This class will serialize an XBlock, producing:
+        (1) A new definition ID for use in Blockstore
+        (2) an XML string defining the XBlock and referencing the IDs of its
+            children (but not containing the actual XML of its children)
+        (3) a list of any static files required by the XBlock and their data
+    """
+
+    def __init__(self, block):
+        """
+        Serialize an XBlock to an OLX string + supporting files, and store the
+        resulting data in this object.
+        """
+        self.orig_block_key = block.scope_ids.usage_id
+        self.static_files = []
+        self.def_id = blockstore_def_key_from_modulestore_usage_key(self.orig_block_key)
+
+        # Create an XML node to hold the exported data
+        olx_node = Element("root")  # The node name doesn't matter: add_xml_to_node will change it
+        # ^ Note: We could pass nsmap=xblock.core.XML_NAMESPACES here, but the
+        # resulting XML namespace attributes don't seem that useful?
+        with override_export_fs(block) as filesystem:  # Needed for XBlocks that inherit XModuleDescriptor
+            # Tell the block to serialize itself as XML/OLX:
+            if not block.has_children:
+                block.add_xml_to_node(olx_node)
+            else:
+                # We don't want the children serialized at this time, because
+                # otherwise we can't tell which files in 'filesystem' belong to
+                # this block and which belong to its children. So, temporarily
+                # disable any children:
+                children = block.children
+                block.children = []
+                block.add_xml_to_node(olx_node)
+                block.children = children
+
+            # Now the block/module may have exported addtional data as files in
+            # 'filesystem'. If so, store them:
+            for item in filesystem.walk():  # pylint: disable=not-callable
+                for unit_file in item.files:
+                    file_path = os.path.join(item.path, unit_file.name)
+                    with filesystem.open(file_path, 'rb') as fh:
+                        data = fh.read()
+                    self.static_files.append(StaticFile(name=unit_file.name, data=data))
+        # Apply some transformations to the OLX:
+        self.transform_olx(olx_node)
+        # Add  <xblock-include /> tags for each child (XBlock XML export
```

Comment:
Were we not going to use xinclude so that the Blockstore does not need to understand OLX specific conventions?

Reply:
Even using this `<xblock-include>` tag, Blockstore still does not need to understand OLX. It's up to the client (runtime) to parse this and then ask Blockstore for the data required to replace the `<xblock-include>` tag with the child OLX.

This comes from https://github.com/open-craft/blockstore/pull/27

Reply:
Ah okay, I had forgotten that!